### ISSUE TITLE
Image segmentation - artifacts on the output classification image

### ISSUE THREAD

**Issue by HardRock4Life** [remicres/otbtf#76]
Hello!

I'm stuck with the image segmentation.

I've preprocessed the images, then extracted patches like this:

![image](https://user-images.githubusercontent.com/21273349/154639446-b0dce785-44e2-46dd-9b73-e4882778584a.png)
![image](https://user-images.githubusercontent.com/21273349/154639516-dd654b59-813f-42ab-ab50-18eaf3205860.png)
![image](https://user-images.githubusercontent.com/21273349/154639555-242ff5bc-d578-4c83-bb7d-89dc4d1801ec.png)
![image](https://user-images.githubusercontent.com/21273349/154639601-1ad05a04-18a9-4572-985e-d6ca53ea2065.png)

The CNN model training runs fine

![image](https://user-images.githubusercontent.com/21273349/154639659-d4725481-7f3b-4939-bca1-f7687cdaa024.png)

For inference, I give a normalized image as as input layer. The final result I get is the following:

![image](https://user-images.githubusercontent.com/21273349/154520970-05760497-b848-43cc-ba42-cd61d4407e3e.png)

![image](https://user-images.githubusercontent.com/21273349/154521036-09a1a75e-922a-4ed4-ace4-41d0108559ce.png)

The way I understand it, the U-NET model is supposed to delete those artifacts in this piece of code?

```
def myModel(x):

  depth = 16
  
  # Encoding
  conv1   = _conv(x,        1*depth)         #  64 x 64 --> 32 x 32 (31 x 31)
  conv2   = _conv(conv1,    2*depth)         #  32 x 32 --> 16 x 16 (15 x 15)
  conv3   = _conv(conv2,    4*depth)         #  16 x 16 -->  8 x  8 ( 7 x  7)
  conv4   = _conv(conv3,    4*depth)         #   8 x  8 -->  4 x  4 ( 3 x  3)
  
  # Decoding (with skip connections)
  deconv1 = _dconv(conv4,           4*depth) #  4  x  4 -->  8 x  8 ( 5 x  5)
  deconv2 = _dconv(deconv1 + conv3, 2*depth) #  8  x  8 --> 16 x 16 ( 9 x  9)
  deconv3 = _dconv(deconv2 + conv2, 1*depth) # 16  x 16 --> 32 x 32 (17 x 17)
  deconv4 = _dconv(deconv3 + conv1, 1*depth) # 32  x 32 --> 64 x 64 (33 x 33)
  
  # Neurons for classes
  estimated = tf.layers.dense(inputs=deconv4, units=nclasses, activation=None)
  
  return estimated
```

Or should it be done differently? Thank you!

**Comment by HardRock4Life** [comment]
It works. 

Solution: Don't forget to delete existing models!
They are not overwriting.

**Comment by remicres** [comment]
Hi @HardRock4Life ,

Overfitting is when the network learns perfectly on the training data, but process poorly on validation data. This can happen when you have not enough training data for instance.
You can check classification metrics of your network on training and validation datasets to see if this is related to overfitting or not (I cannot tell from the provided illustration if it is the case).

Generally, the more training samples you have, the best is the result (if your have quality data). The tutorial in the book is an insight to semantic segmentation, and in real life you will definitely seek for large amount of training data to train the best model.

Maybe the model has mapped road as "buildings" because it hasn't seen enough training examples containing roads?
